### PR TITLE
Preserve headers in multipart stream uploads

### LINKED ISSUE #472: Streaming multipart uploads drop custom headers at CHUNK_SIZE
**Describe the bug**

`put_object_stream_builder` applies per-upload metadata and custom headers when an upload uses single-part `PutObject`. When the payload reaches `CHUNK_SIZE` and switches to multipart upload, those headers are not sent on `CreateMultipartUpload`.

S3-compatible services require completed-object configuration such as user metadata, cache control, storage class, and SSE-S3/KMS headers during multipart initiation. The upload succeeds, but the completed object silently loses that configuration.

**To Reproduce**

1. Start an S3-compatible MinIO instance and create a bucket.
2. Upload synthetic byte payloads using `put_object_stream_builder` with custom metadata.
3. Check the completed object's metadata with `HEAD`.

| Payload size | Upload path | Metadata |
| ---: | --- | --- |
| `CHUNK_SIZE - 1` | single-part | present |
| `CHUNK_SIZE` | multipart | missing |
| above `CHUNK_SIZE` | multipart | missing |

**Expected behavior**

Streaming-builder headers should be signed and sent on `CreateMultipartUpload`, so completed-object metadata and other initiation-time configuration survive at and above the multipart boundary.

**Environment**

- Rust: 1.97.1
- rust-s3: 0.37.2
- Storage: MinIO-compatible S3 endpoint

**Additional context**

This is the multipart-specific gap related to #408. A previous implementation direction was discussed in #446. The consumer-visible reproduction is brunojppb/turbo-cache-server#620, and the proposed fix is #471.

### PR DESCRIPTION
**Pull request by mfroembgen**
## Problem

`put_object_stream_builder` accepts per-upload metadata and headers, but drops them when the payload reaches `CHUNK_SIZE` and switches to multipart upload. S3 requires completed-object headers such as metadata, cache control, storage class, and SSE-S3/KMS configuration on `CreateMultipartUpload`.

## Solution

- add a backward-compatible `initiate_multipart_upload_with_headers` API while preserving the existing method signature
- forward streaming-builder headers to multipart initiation through a request-scoped bucket clone
- keep multipart parts, completion, bounded concurrency, and error handling unchanged
- add a hermetic request/signing regression plus real MinIO coverage below, at, and above `CHUNK_SIZE`

The request-scoped clone preserves the existing credential provider and client configuration without mutating shared bucket headers.

Fixes #472.

This follows the maintainer direction in #446 and is related to #408. The motivating consumer bug is brunojppb/turbo-cache-server#620.

## Verification

- `cargo fmt --all -- --check`
- `cargo test -p rust-s3 multipart_initiation_with_headers_sends_and_signs_object_configuration`
- sync API compile with `sync-native-tls`
- Tokio and async-std MinIO regression at 8,388,607, 8,388,608, and 8,715,039 bytes
- MinIO HEAD metadata and downloaded object bytes verified at every boundary size

The request test also verifies that metadata, cache control, and SSE-S3 are included in SigV4 `SignedHeaders` on multipart initiation.

## Summary by CodeRabbit

* **New Features**
  * Added support for custom headers when initiating multipart uploads.
  * Streaming multipart uploads now preserve custom headers throughout the upload process.
  * Existing multipart upload methods remain available without requiring header configuration.

* **Bug Fixes**
  * Improved header transmission and request signing across multipart upload size boundaries.
  * Preserved metadata consistently during multipart uploads.

- - -
This change is [Reviewable](https://reviewable.io/reviews/durch/rust-s3/471)

### CODE REVIEW

**Comment by ghost** [comment]
### Brownian Motion (Brass)
  
  **Recommendation:** Refactor
  
  **Summary:** Fixes lost multipart headers; good direction, but watch API bloat and async test certainty
  **Risk:** Medium   ·   **Confidence:** 78%
  
  **Highlights**
- Targets a real S3 multipart semantics gap
- Adds regression tests for request signing and MinIO boundaries
- Preserves existing API via wrapper method

**Unknowns**
- I can't see the implementation of RequestImpl::new / header merge order
- It's unclear whether the new tests are stable across all enabled backends in CI
- The linked issue is closed, but I can't confirm whether this PR matches the maintainer's preferred public API shape

**Next actions**
- Keep: forwarding headers only on multipart initiation; that is the actual fix
- Drop: any further API proliferation unless there is a concrete consumer need
- Add: one focused compatibility test proving single-part uploads still behave exactly as before
- Add: explicit docs or changelog note that these headers affect multipart initiation, not parts
- Add: verify whether extra_headers cloning preserves/overrides header precedence exactly as intended

**Reflection questions**
- Why is a new public method the right abstraction instead of extending the existing one with an optional parameter everywhere?
- Do we actually need a request-scoped bucket clone, or can we pass initiation headers more directly without widening the surface area?
- What breaks if a caller expects the bucket's shared extra_headers to remain the single source of truth?

**Comment by ghost** [comment]
### Brownian Motion (Brass)
  
  **Recommendation:** Refactor
  
  **Summary:** Fixes header loss at multipart initiation; approach is sound but needs tighter validation
  **Risk:** Medium   ·   **Confidence:** 78%
  
  **Highlights**
- Addresses a real S3 multipart semantics bug
- Keeps existing API intact while adding a scoped opt-in path
- Adds hermetic and MinIO regression coverage

**Unknowns**
- Whether any other multipart initiation paths bypass this new helper
- Whether all documented headers are supported consistently across AWS, MinIO, and other S3-compatible backends
- Whether the added public API is worth long-term maintenance versus extending the existing builder

**Next actions**
- Keep: the request-scoped bucket clone to avoid mutating shared headers
- Keep: multipart initiation regression tests around CHUNK_SIZE
- Drop: duplicated sync/async documentation and example churn unless strictly needed
- Add: a focused test proving header precedence/merge rules for duplicate keys
- Add: confirmation that all multipart-object headers are actually accepted by the supported backends, not just signed

**Reflection questions**
- What core assumption underpins this PR's approach: that header preservation belongs at initiation, not later?
- Why is a new public initiate_multipart_upload_with_headers API necessary instead of plumbing the existing builder state internally?
- Could the same fix be achieved with less API surface and less duplicated doc/test code?

File: s3/src/bucket.rs [hunk 1]
```diff
@@ -1866,8 +1866,55 @@ impl Bucket {
...
+    pub async fn initiate_multipart_upload_with_headers(
+        &self,
+        s3_path: &str,
+        content_type: &str,
+        custom_headers: Option<HeaderMap>,
+    ) -> Result<InitiateMultipartUploadResponse, S3Error> {
+        let mut request_bucket = self.clone();
+        if let Some(custom_headers) = custom_headers {
+            request_bucket.extra_headers.extend(custom_headers);
+        }
```
[the remark this answers is not in the document]
Comment (mfroembgen) [reply s3/src/bucket.rs:1915]: Verified against the `http` 1.x implementation used here: `HeaderMap::extend(HeaderMap)` replaces the destination values for the first value of each incoming key via `Occupied::insert`, then preserves any additional values from the incoming map. It therefore does not retain the bucket-level value in this case.

I kept the merge implementation unchanged and added a regression in 97f98d0 that starts with `x-amz-meta-artifact-tag: bucket-default` at bucket scope, supplies `synthetic-signature` at request scope, and verifies that only the request-scoped value is sent. The focused TCP test passes. The same follow-up also adds the requested sync API documentation example.
